File: bstvalidate.go
// Time Complexity : O(n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this :

// Your code here along with comments explaining your approach
/*
call recursive function with root, min as nil, max as nil

in recursive function
if root is nil then return true
if min is not nil and root value<= min value return false
if max is not nil and root value>= max value return false
return recursive(root.left, min, root) && recursive(roo.right, root, max)
*/
package main

import "fmt"

type TreeNode struct {
     Val int
     Left *TreeNode
     Right *TreeNode
}

var prev *TreeNode
func isValidBST(root *TreeNode) bool {
	return inorder(root, nil, nil)
}

func inorder(root, min, max *TreeNode) bool {
	if root == nil {
		return true
	}
	if min != nil && root.Val <= min.Val {
		return false
	}
	if max != nil && root.Val >= max.Val {
		return false
	}
	return inorder(root.Left, min, root) && inorder(root.Right, root, max)
}

func MainValidBST() {
	root := &TreeNode{
		Val: 0,
		Left: nil,
		Right: nil,
	}
	fmt.Println(isValidBST(root)) // expected true
}

File: bst_validate.py
# // Time Complexity :O(n)
# // Space Complexity :O(h)
# // Did this code successfully run on Leetcode :yes
# // Any problem you faced while coding this :no


# // Your code here along with comments explaining your approach







# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def __init__(self):
        self.max=-9999999999999
        self.flag=0
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        self.inorder(root)
        if self.flag==1:
            return False
        else:
            return True
    def inorder(self,root):
        if root:
            self.inorder(root.left)
            print(root.val)
            print(self.max)
            if self.max>=root.val:
                self.flag=1
                return 
            else:
                self.max=root.val
            print(root.val)
            self.inorder(root.right)
        else:
            return
        